src/compiler/checker.ts

```diff
@@ -42772,6 +42790,22 @@ namespace ts {
             return false;
         }
 
+        function checkGrammarImportCallArguments(node: ImportCall, nodeArguments: NodeArray<Expression>): boolean {
+            if (moduleKind !== ModuleKind.ESNext) {
+                // We are allowed trailing comma after proposal-import-assertions.
+                checkGrammarForDisallowedTrailingComma(nodeArguments);
+
+                if (nodeArguments.length > 1) {
+                    const assertionArgument = nodeArguments[1];
+                    return grammarErrorOnNode(assertionArgument, Diagnostics.Dynamic_import_only_supports_a_second_argument_when_the_module_option_is_set_to_esnext);
+                }
+            }
+            if (nodeArguments.length !== 1) {

```

Comment:
This condition is what's causing the errors in https://github.com/microsoft/TypeScript/pull/40698/files#r705675967

- This should probably be in an `else` branch (so that it only applies when `moduleKind` is `ESNext`
- This should probably be checking `nodeArguments.length === 0 || nodeArguments.length > 2`